crates/libtest2/src/macros.rs

```diff
@@ -99,3 +121,80 @@ macro_rules! _parse_ignore {
         $context.ignore()
     };
 }
+
+#[macro_export]
+macro_rules! _run_test {
+    ($context:expr, [$($expected:literal)?]) => {
+        $crate::assert_panic!(run($context), $($expected)?)
+    };
+    ($context:expr $(,)?) => {{
+        run($context)
+    }};
+}
+
+#[macro_export]
+macro_rules! assert_panic {
```

Comment:
How much of this can live in a function?  This is a lot of code to put in a macro which can negatively impact compile times

Reply:
If `assert_panic` is an implementation detail of the macro, then I don't mind there being one variant.  The runtime overhead of `Option` should be minimal.

If it's not, then I would go with `assert_panic` and  `assert_panic_message` (don't need `with` because we're asserting on the message, not panicking itself) / `assert_panic_contains` (express what message operation we do).

One thing in favor of a macro is adding a display message like other assert macros ... but that would be confusing with the panic message being optional.